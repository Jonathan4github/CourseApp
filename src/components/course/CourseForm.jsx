import React from 'react';
import PropTypes from 'prop-types';

import TextInput from '../common/TextInput';
import SelectInput from '../common/SelectInput';


const propTypes = {
  course: PropTypes.shape().isRequired,
  allAuthors: PropTypes.arrayOf(PropTypes.object).isRequired,
  handleOnSave: PropTypes.func.isRequired,
  handleOnChange: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  errors: PropTypes.shape()
};

const defaultProps = {
  errors: {}
};

const CourseForm = ({
  course, allAuthors, handleOnSave, handleOnChange, isLoading, errors
}) => (
  <form>
    <TextInput
      name="title"
      label="Title"
      value={course.title}
      onChange={handleOnChange}
      error={errors.title} />
    <SelectInput
      name="authorId"
      label="Author"
      value={course.authorId}
      defaultOption="Select Author"
      options={allAuthors}
      onChange={handleOnChange}
      error={errors.authorId} />
    <TextInput
      name="category"
      label="category"
      value={course.category}
      onChange={handleOnChange}
      error={errors.category} />
    <TextInput
        name="length"
        label="length"
        value={course.length}
        onChange={handleOnChange}
        error={errors.length} />
    <input
     type="submit"
     className="btn btn-primary"
     value={isLoading ? 'Saving' : 'Save'}
     onClick={handleOnSave} />
  </form>
);

CourseForm.propTypes = propTypes;
CourseForm.defaultProps = defaultProps;

export default CourseForm;
