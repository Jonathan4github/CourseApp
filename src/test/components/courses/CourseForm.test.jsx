/* eslint-disable */
import CourseForm from '../../../components/course/CourseForm';
import courseMock from '../../__mocks__/courseMock';


describe('<CourseForm />', () => {
  const setup = () => {
    const props = {
      course: courseMock.existingCourse,
      handleOnSave: jest.fn(),
      handleOnChange: jest.fn(),
      handleOnFocus: jest.fn(),
      isSaving: false,
      errors: {}
    };

    return {
      shallowWrapper: shallow(<CourseForm {...props} />),
      props
    };
  };

  it('renders component without crashing', () => {
    const { shallowWrapper } = setup();

    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('should call onChange() event', () => {
    const { shallowWrapper, props } = setup();
    shallowWrapper.setProps({
      ...props,
      isSaving: true
    });

    expect(shallowWrapper.find('input').props().value).toEqual('Saving...');
  });

  it('renders three(3) <TextInput /> child components', () => {
    const { shallowWrapper } = setup();
    expect(shallowWrapper.find('TextInput').length).toBe(3);
  });

  it('calls handleOnSave method when form is submitted', () => {
    const { shallowWrapper, props } = setup();

    shallowWrapper.find('input').simulate('click');
    expect(props.handleOnSave).toHaveBeenCalled();
  });
});
