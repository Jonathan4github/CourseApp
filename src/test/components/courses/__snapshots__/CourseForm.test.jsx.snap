// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CourseForm /> renders component without crashing 1`] = `
<form>
  <TextInput
    error=""
    label="Title"
    name="title"
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    placeholder="eg. Android Fundamentals..."
    value="Building Applications in React and Flux"
  />
  <SelectInput
    defaultOption="Select Author"
    error=""
    label="Author"
    name="authorId"
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    options={Array []}
    value="cory-house"
  />
  <TextInput
    error=""
    label="category"
    name="category"
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    placeholder="eg. Mobile, UI/UX..."
    value="JavaScript"
  />
  <TextInput
    error=""
    label="length"
    name="length"
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    placeholder="eg. 4:59"
    value="5:08"
  />
  <input
    className="btn btn-primary"
    disabled={false}
    onClick={[MockFunction]}
    type="submit"
    value="Save"
  />
</form>
`;
